zio-http/shared/src/main/scala/zio/http/codec/HttpContentCodec.scala

```diff
@@ -0,0 +1,195 @@
+package zio.http.codec
+
+import scala.collection.immutable.ListMap
+
+import zio._
+
+import zio.stream.ZPipeline
+
+import zio.schema.Schema
+import zio.schema.codec._
+
+import zio.http._
+import zio.http.codec.HttpContentCodec.json.fromSchema
+
+final case class HttpContentCodec[A](
+  choices: ListMap[MediaType, BinaryCodec[A]],
+  schema: Schema[A],
+) { self =>
+
+  /**
+   * A right biased merge of two HttpContentCodecs.
+   */
+  def ++(that: HttpContentCodec[A]): HttpContentCodec[A] =
+    HttpContentCodec(choices ++ that.choices, schema)
+
+  def decodeRequest(request: Request): Task[A]    = {
+    val contentType =
+      request.header(Header.ContentType).getOrElse(Header.ContentType(MediaType.application.`json`))
+    lookup(contentType) match {
+      case Some(codec) =>
+        request.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+  def decodeResponse(response: Response): Task[A] = {
+    val contentType =
+      response.header(Header.ContentType).getOrElse(Header.ContentType(MediaType.application.`json`))
+    lookup(contentType) match {
+      case Some(codec) =>
+        response.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+
+  def encode(value: A): Either[String, Body] =
+    choices.headOption match {
+      case Some((mediaType, codec)) =>
+        Right(Body.fromChunk(codec.encode(value)).contentType(mediaType))
+      case None                     =>
+        Left("No codec found")
+    }
+
+  def only(mediaType: MediaType): HttpContentCodec[A] =
+    HttpContentCodec(
+      ListMap(
+        mediaType -> lookup(mediaType)
+          .getOrElse(
+            throw new IllegalArgumentException(s"MediaType $mediaType is not supported by $self"),
+          ),
+      ),
+      schema,
+    )
+
+  def only(mediaType: Option[MediaType]): HttpContentCodec[A] = {
+    mediaType match {
+      case Some(mediaType) =>
+        HttpContentCodec(
+          ListMap(
+            mediaType -> lookup(mediaType)
+              .getOrElse(
+                throw new IllegalArgumentException(s"MediaType $mediaType is not supported by $self"),
+              ),
+          ),
+          schema,
+        )
+      case None            =>
+        self
+    }
+  }
+
+  def onlyOrNone(mediaType: MediaType): Option[HttpContentCodec[A]] =
+    if (choices.contains(mediaType)) Some(only(mediaType))
+    else None
+
+  def lookup(mediaType: MediaType): Option[BinaryCodec[A]] =
```

Comment:
Due to extreme overhead, these `lookup` methods cannot be executed on a hot path.

I would suggest to precompute a map, stored as a private member of `HttpContentCodec`, which allows you to quickly perform the `matches` comparison against a given `MediaType`, and which already returns an `Option` (preallocated `Some`), to avoid all this overhead.

Then, of course, we must say: no dynamic `HttpContentCodec`, because it will be more expensive to create them. But that's fine, we would want to do that anyway.

Reply:
I think precomputing all is too expensive. Imagine you have to create all `text/<whatever>` keys in the map.  And then never use it. There are almost 2k predefined media types (all lazy val) in zio-http.
I would implement a cache via `ConcurrentHashMap`?

Reply:
Sure!

Reply:
Btw you don't need concurrent hash map. Just a `var cache: Map[...]`. Faster